packages/core-js/modules/web.url-search-params.constructor.js

```diff
@@ -57,31 +58,139 @@ var split = uncurryThis(''.split);
+var fromCodePoint = getBuiltIn('String', 'fromCodePoint');
+var $parseInt = parseInt;
+
+var parseHexOctet = function (string, start) {
+  var substr = stringSlice(string, start, start + 2);
+  if (!VALID_HEX.test(substr)) return NaN;
+
+  return $parseInt(substr, 16);
 };
 
-var percentDecode = function (sequence) {
-  try {
-    return decodeURIComponent(sequence);
-  } catch (error) {
-    return sequence;
+var getLeadingOnes = function (octet) {
+  var binString = padStart(numberToString(octet, 2), 8, '0');
+  return indexOf(binString, '0') !== -1 ? indexOf(binString, '0') : binString.length;
+};
+
+var utf8Decode = function (octets) {
+  var len = octets.length;
+  var codePoint = null;
+
+  switch (len) {
+    case 1:
+      codePoint = octets[0];
+      break;
+    case 2:
+      codePoint = (octets[0] & 0x1F) << 6 | (octets[1] & 0x3F);
+      break;
+    case 3:
+      codePoint = (octets[0] & 0x0F) << 12 | (octets[1] & 0x3F) << 6 | (octets[2] & 0x3F);
+      break;
+    case 4:
+      codePoint = (octets[0] & 0x07) << 18 | (octets[1] & 0x3F) << 12 | (octets[2] & 0x3F) << 6 | (octets[3] & 0x3F);
+      break;
   }
+
+  return codePoint > 0x10FFFF ? null : codePoint;
 };
 
-var deserialize = function (it) {
-  var result = replace(it, plus, ' ');
-  var bytes = 4;
-  try {
-    return decodeURIComponent(result);
-  } catch (error) {
-    while (bytes) {
-      result = replace(result, percentSequence(bytes--), percentDecode);
+/* eslint-disable max-statements -- TODO */
+var decode = function (input) {
+  var length = input.length;
+  var result = '';
+  var i = 0;
+
+  while (i < length) {
+    var decodedChar = input[i];
+
+    if (decodedChar === '%') {
+      if (i + 3 > length && i + 1 !== length) {
+        /* eslint-disable no-useless-assignment -- TODO */
+        decodedChar = FALLBACK_REPLACER;
+        break;
+      }
+
+      if (input[i + 1] === '%' || i + 1 === length) {
+        result += '%';
+        i++;
+        continue;
+      }
+
+      var octet = parseHexOctet(input, i + 1);
+
+      if (isNaN(octet)) {
+        result += decodedChar;
+        i++;
+        continue;
+      }
+
+      i += 2;
+      var byteSequenceLength = getLeadingOnes(octet);
+
+      if (byteSequenceLength === 0) {
+        decodedChar = fromCharCode(octet);
+      } else {
+        if (byteSequenceLength === 1 || byteSequenceLength > 4) {
+          result += FALLBACK_REPLACER;
+          i++;
+          continue;
+        }
+
+        var octets = [octet];
+        var sequenceIndex = 1;
+
+        while (sequenceIndex < byteSequenceLength) {
+          i++;
+          if (i + 3 > length || input[i] !== '%') {
+            break;
+          }
+
+          var nextByte = parseHexOctet(input, i + 1);
+
+          if (nextByte > 191 || nextByte < 128) {  // incorrect next byte
+            break;
+          }
+
+          if (isNaN(nextByte)) {
+            i += 3;
+            break;
+          }
+          octets.push(nextByte);
```

Comment:
Don't use methods directly from instance, cache it and unbind with `uncurryThis`.